### PR TITLE
Create new abstract metadata table from sync rules

### PR DESCRIPTION
**Pull request by SferaDev**
### :pushpin: References

- This is a refactor or #109 to allow re-using the new component

### :memo: Implementation

- Abstract metadata table from Sync Rules and re-use it in the all the Sync Pages

- This allows us to include Groups with ease see #117 and #118 

### :fire: Is there anything the reviewer should know to test it?

- You can test either the specific synchronization pages (OU, DE, ...) or the SyncRules metadata selection step. All of them are using this new abstracted component.

### CODE REVIEW

File: src/models/syncRule.ts [hunk 1]
```diff
@@ -128,6 +91,43 @@ export default class SyncRule {
...
+                    ...this.syncRule.builder.metadata,
+                    [type]: updated,
+                },
+            },
+        });
+    }
+
+    public removeMetadataIds(ids: string[]): SyncRule {
+        const metadata = _.clone(this.syncRule.builder.metadata);
+        for (const type in metadata) _.pullAll(metadata[type], ids);
```
Comment (tokland) [inline src/models/syncRule.ts]: `_.pullAll` mutates the input (thus the `_.clone`), proposal (untested!):

```
const metadata = _(this.syncRule.builder.metadata)
            .mapValues(idsForType => _.difference(idsForType, ids))
            .value();
```

**Comment by SferaDev** [comment]
Fixed the comments from @tokland

**Comment by adrianq** [comment]
@SferaDev There are a few warnings in the js console:
image

and most importantly, a critical error when entering the sync rules section:
image

**Comment by adrianq** [comment]
@SferaDev The problem happens when there is a sync rule for a target instance that has been deleted from the list of configured instances

**Comment by SferaDev** [comment]
> @SferaDev The problem happens when there is a sync rule for a target instance that has been deleted from the list of configured instances

Yes, fixed in #120

**Comment by SferaDev** [comment]
> @SferaDev There are a few warnings in the js console:
> image
> 
> and most importantly, a critical error when entering the sync rules section:
> image

Will take a look on these tomorrow.

**Comment by adrianq** [comment]
Great, merging this!

**Pull request merged**